dl/callbacks.py

```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
 
         self.time = time.time()
         state.step += bs
-        state.batch_metrics = defaultdict(lambda: 0)
+        #
 
     def on_loader_end(self, state):
         lm = state.loader_mode
 
         state.epoch_metrics[lm] = {
             key: UtilsFactory.get_val_from_metric(value)
-            for key, value in state.epoch_metrics[lm].items()}
+            for key, value in state.epoch_metrics[lm].items()
+        }
 
         for key, value in state.epoch_metrics[lm].items():
             self.loggers[lm].add_scalar(f"epoch {key}", value, state.epoch)
 
         epoch_metrics_str = "\t".join([
             "{key} {value:.4f}".format(key=key, value=value)
-            for key, value in sorted(state.epoch_metrics[lm].items())])
+            for key, value in sorted(state.epoch_metrics[lm].items())
+        ])
 
-        print("{epoch} * Epoch ({mode}): {metrics}".format(
-            epoch=state.epoch, mode=lm, metrics=epoch_metrics_str))
+        if self.verbose:
+            print("{epoch} * Epoch ({mode}): {metrics}".format(
+                epoch=state.epoch, mode=lm, metrics=epoch_metrics_str))
 
         if self.reset_step:
             state.step = None
 
 
+class Logger(Callback):
+    def __init__(self, log_dir):
+        super().__init__()
+        os.makedirs(log_dir, exist_ok=True)
+        log_filepath = os.path.join(log_dir, 'n01_logs.txt')
+        self.logger = self._get_logger(log_filepath)
+
+    def on_epoch_begin(self, state):
+        self.logger.info('Epoch {} | optimizer "{}" | lr {}'.format(
+            state.epoch, state.optimizer.__class__.__name__,
+            self._get_current_lr(state)))
+
+    def on_epoch_end(self, state):
+        self.logger.info("Train metrics: " + self._get_metrics_string(
+            state.epoch_metrics['train']))
+        self.logger.info("Valid metrics: " + self._get_metrics_string(
+            state.epoch_metrics['valid']) + "\n")
+
+    def on_train_begin(self, state):
+        self.logger.info(
+            'Starting training with params:\n{}\n\n'.format(state))
+
+    def on_batch_end(self, state):
+        if state.is_train:
```

Comment:
why only `is_train`?

Reply:
Because in validation mode i'm intersected only in total metrics